grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerProvider.java

```diff
@@ -0,0 +1,99 @@
+package org.hypertrace.circuitbreaker.grpcutils.resilience;
+
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerConfig;
+import io.github.resilience4j.circuitbreaker.CircuitBreakerRegistry;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerConfigParser;
+
+/** Utility class to provide Resilience4j CircuitBreaker */
+@Slf4j
+class ResilienceCircuitBreakerProvider {
+
+  private static final String SHARED_KEY = "SHARED_KEY";
+  private final CircuitBreakerRegistry circuitBreakerRegistry;
+  private final Map<String, CircuitBreakerConfig> circuitBreakerConfigMap;
+  private final Map<String, CircuitBreaker> circuitBreakerCache = new ConcurrentHashMap<>();
+  private final List<String> disabledKeys;
+  private final boolean defaultEnabled;
+
+  public ResilienceCircuitBreakerProvider(
+      CircuitBreakerRegistry circuitBreakerRegistry,
+      Map<String, CircuitBreakerConfig> circuitBreakerConfigMap,
+      List<String> disabledKeys) {
+    this.circuitBreakerRegistry = circuitBreakerRegistry;
+    this.circuitBreakerConfigMap = circuitBreakerConfigMap;
+    this.disabledKeys = disabledKeys;
+    this.defaultEnabled = !disabledKeys.contains(CircuitBreakerConfigParser.DEFAULT_THRESHOLDS);
+  }
+
+  public Optional<CircuitBreaker> getCircuitBreaker(String circuitBreakerKey) {
+    if (disabledKeys.contains(circuitBreakerKey)) {
+      return Optional.empty();
+    }
+    return Optional.ofNullable(
+        circuitBreakerCache.computeIfAbsent(
+            circuitBreakerKey,
+            key -> {
+              CircuitBreaker circuitBreaker =
+                  getCircuitBreakerFromConfigMap(circuitBreakerKey, defaultEnabled);
+              // If no circuit breaker is created return empty
+              if (circuitBreaker == null) {
+                return null; // Ensures cache does not store null entries
+              }
+              attachListeners(circuitBreaker);
+              return circuitBreaker;
+            }));
+  }
+
+  public Optional<CircuitBreaker> getSharedCircuitBreaker() {
+    if (!defaultEnabled) {
+      return Optional.empty();
+    }
+    return Optional.of(
```

Comment:
nit - can this just be `return getCircuitBreaker(SHARED_KEY);`?

Reply:
resolved